### PR TITLE
feat: Add remote invoke implementation for Kinesis stream service for put_record API

### PR DESCRIPTION
**Pull request by hnnasit**
#### Which issue(s) does this change fix?

N/A

#### Why is this change necessary?
This PR adds the implementation for remote invoke support for Kinesis stream service. The boto3 API [put_record](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/kinesis/client/put_record.html) is called when sam remote invoke is run to put a record to the provided kinesis stream name using event or event-file.

#### How does it address the issue?
- The `Data` of the record which is a required field can be provided using `event` or `event-file` parameters which will get encoded as base64 and put into Kinesis stream. If no event or event-file is provided, `{}` is used as `Data` for putting a record. The event is loaded using json before passing to boto3 API and the event-file is read and directly passed to boto3 API.
- The `PartitionKey` is a required field for boto3 API and so it is randomly generated if not provided as input. This is just used to determine which Shard does the record get added to using a MD5 hash function.
- The other fields such as `ExplicitHashKey`, `SequenceNumberForOrdering`, and `StreamARN` can be provided as strings using the `--parameter` field

Use cases:
1. `sam remote invoke KinesisStreamLogicalId --stack-name sam-app --event '{"hello": "world", "foo": 1, "bar": {}}'`: Puts a record with provided event as data to the `KinesisStreamLogicalId` Kinesis stream.
```
{
  "ShardId": "shardId-000000000000",
  "SequenceNumber": "sequence-number"
}
```
3. `sam remote invoke arn:aws:kinesis:us-east-1:12345678910:stream/kinesis-stream-name --event "hello world"`: Kinesis stream Arn used as resource-id to put a record
4. `sam remote invoke kinesis-stream-name --event "hello world" --output json`: Kinesis stream name used as resource-id to put a record
```
{
  "ShardId": "shardId-000000000000",
  "SequenceNumber": "sequence-number",
  "ResponseMetadata": {}
}
```

#### What side effects does this change have?
No side effects.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/remote_invoke/kinesis_invoke_executors.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+@dataclass
+class KinesisStreamPutRecordTextOutput:
+    """
+    Dataclass that stores put_record boto3 API fields used to create
+    text output.
+    """
+
+    ShardId: str
+    SequenceNumber: str
+
```
Comment (mndeveci) [inline samcli/lib/remote_invoke/kinesis_invoke_executors.py:38]: Should these be named as `shard_id` and `sequence_number`?

File: requirements/reproducible-mac.txt [hunk 2]
```diff
@@ -415,6 +415,10 @@ mypy-boto3-iam==1.28.37 \
     --hash=sha256:39bd5b8b9a48cb47d909d45c13c713c099c2f84719612a0a848d7a0497c6fcf4 \
     --hash=sha256:a5ed8c70c610f2ae7ee4a32ecf87c42824bfeb1e8a8c6347d888c73383d2c61f
     # via boto3-stubs
+mypy-boto3-kinesis==1.28.36 \
```
Comment (mndeveci) [inline requirements/reproducible-mac.txt:418]: Will these be updated by the other PR that you raised?

Comment (hnnasit) [reply requirements/reproducible-mac.txt:418]: Yes, I will merge that PR first and update this PR.

File: samcli/lib/remote_invoke/kinesis_invoke_executors.py (hunk 1 above)
Comment (hnnasit) [reply samcli/lib/remote_invoke/kinesis_invoke_executors.py:38]: I wanted to name them with an `_` as well but the boto3 API returns the values `SequenceNumber` and `ShardId` and so I thought we should keep the remote invoke output consistent with the API returned response names.

**Review by mndeveci — APPROVED** [review]

File: samcli/commands/remote/remote_invoke_context.py [hunk 3]
```diff
@@ -215,6 +215,11 @@ def _get_from_physical_resource_id(self) -> CloudFormationResourceSummary:
                 sqs_client = self._boto_client_provider("sqs")
                 resource_id = get_queue_url_from_arn(sqs_client, resource_arn.resource_id)
 
+            if SUPPORTED_SERVICES.get(service_from_arn) == AWS_KINESIS_STREAM:
+                # Note (hnnasit): Add unit test after AWS_KINESIS_STREAM is added to SUPPORTED_SERVICES
+                # StreamName extraced from arn is used as resource_id.
```
Comment (hawflau) [inline samcli/commands/remote/remote_invoke_context.py:220]: nit
```suggestion
                # StreamName extracted from arn is used as resource_id.
```

**Review by hawflau — APPROVED** [review]
LGTM. just one minor typo

**Pull request merged**